### PR TITLE
Find more looping implicits

### LINKED ISSUE #10947: infinite loop when evaluating a recursive given definition
When running a program which consumes opaque types as follows, it appears to be stuck on an infinite loop. Not sure what is going on here.

## Minimized code

```Scala
object Prices {
  opaque type Price = BigDecimal

  object Price{
    given Ordering[Price] = summon[Ordering[BigDecimal]]
  }

}

extension[K,V] (m1: Map[K,V]) def computeDiff(m2: Map[K,V])(using Ordering[K])(using orderAsc: Boolean = true): Unit = println("Ok....")

import Prices._
@main def mainz = {
    println("begin")
    val result = Map.empty[Price,Int].computeDiff(Map.empty[Price,Int])
    println(s"end $result")
}
```

Scastie link: https://scastie.scala-lang.org/ajEagvzvTbitmzbqFt0Oxw

## Output

```scala
begin
```

## Expectation
Should successfully print the following:
```scala
begin
Ok....
end
```

This works if the `Price` was swapped with `BigDecimal` in the line
```scala
  val result = Map.empty[Price,Int].computeDiff(Map.empty[Price,Int])
```

This is happening on 3.0.0-M3

### LINKED ISSUE #15474: inside Conversion methods, resolve extension wrong
## Compiler version

3.1.2

## Demo1 1

```Scala
  given Conversion[ String, Int ] with
    def apply(from: String): Int = from.toInt   // the generate code is def apply(from:String:Int) = apply(from)
```

### Output

The generated code is:
```scala
 public int apply(java.lang.String);
    Code:
       0: aload_0
       1: aload_1
       2: invokevirtual #34                 // Method apply:(Ljava/lang/String;)I, calls apply recusive, so will StackOverflow.
       5: ireturn
```

## Demo2
```scala
object Demo2:
  def arr2list[T](arr: Array[T]): List[T] = arr.toList

  given [T]: Conversion[Array[T], List[T]] with
    def apply(arr: Array[T]): List[T] = arr2list(arr)

  def main(args: Array[String]): Unit = 
    val arr = Array(1,2,3)
    val list: List[Int] = arr
    println(s"list = ${list}")
  
```

The code still compiled to 
```scala
object Demo2:
  def arr2list[T](arr: Array[T]): List[T] = given_Conversion_Array_List.apply(arr) // instead arr.toList

  given [T]: Conversion[Array[T], List[T]] with
    def apply(arr: Array[T]): List[T] = arr2list(arr)
```

so run the code will got the StackOverflow error.

### PR DESCRIPTION
**Pull request by odersky**
1. Also check apply methods of companions of implicit or given classes.
    Specifically apply methods of implicit Conversions.

 2. Look inside Inlined nodes to detect loops.

Fixes #15474
Fixes #10947

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/CheckLoopingImplicits.scala [hunk 1]
```diff
@@ -84,7 +87,9 @@ class CheckLoopingImplicits extends MiniPhase:
         checkNotLooping(t.rhs)
       case _ =>
 
-    if sym.isOneOf(GivenOrImplicit | Lazy | ExtensionMethod) then
+    if sym.isOneOf(GivenOrImplicit | Lazy | ExtensionMethod)
+      || sym.name == nme.apply && sym.owner.is(Module) && sym.owner.sourceModule.isOneOf(GivenOrImplicit)
```
Comment (julienrf) [inline compiler/src/dotty/tools/dotc/transform/CheckLoopingImplicits.scala:91]: Could you please add a comment to explain the intent behind the condition on the right side of the `||`? I wonder if this condition is too strong, but I am not sure. Do we have existing “pos” tests?

**Comment by som-snytt** [comment]
This will help folks who need help and also save time on tickets!

**Comment by wangzaixiang** [comment]
I tried the main branch for the following code (build from commit: de3a82c0c6c1f04dd8c6305c0098d1988de3774a, Find more looping implicits):

```scala
object Demo2: 

  given Conversion[String, Int] with
    def apply(s: String): Int = s.toInt

  def main(args: Array[String]): Unit = 
    val s = "123"
    val i: Int = s
    println(s"s=$s i=$i }")

```

It runs with StackOverflow still.

**Comment by som-snytt** [comment]
@wangzaixiang Your example is now a test. Note that `s.toInt` requires a conversion, and it uses the conversion defined here (not the "usual" one), so that it loops. (Your ticket called it wrong, but it is expected; the linked tickets have more examples.) I'm not sure whether you approve this resolution.

**Comment by wangzaixiang** [comment]
> @wangzaixiang Your example is now a test. Note that `s.toInt` requires a conversion, and it uses the conversion defined here (not the "usual" one), so that it loops. (Your ticket called it wrong, but it is expected; the linked tickets have more examples.) I'm not sure whether you approve this resolution.

Sure. I got it.   It's not a BUG.

**Comment by som-snytt** [comment]
My eyes went briefly out of focus when I looked at the example. It's like an optical illusion.

To clarify further, the conversion to `Int` is chosen because `Int` does have a `toInt` method.

Rubbing my eyes.

**Comment by WojciechMazur** [comment]
@odersky Isn't needed change to check file of test i13044 a regression? Since introducing this change Caliban is not able to build in the community build (it's using -Xfatal-warnings) 
Test i13044 was a reproduction of issues Caliban found when upgrading 3.0.0 -> 3.0.1
3.2.1-RC2-nightly logs for failing Caliban build https://scala3.westeurope.cloudapp.azure.com/blue/rest/organizations/jenkins/pipelines/buildCommunityProject/runs/890/nodes/12/log/?start=0

The mentioned test file is used 2 times: 
* in neg tests where it fails and output is being checked
* in pos tests with "-Xmax-inlines:33", which it still passes, but it's warnings are being ignored

**Comment by ghostdogpr** [comment]
The code that fails to compile is deriving an instance for a recursive ADT (simplified):
```scala
sealed trait InputValue
object InputValue {
  case class ListValue(values: List[InputValue])          extends InputValue 
  case class ObjectValue(fields: Map[String, InputValue]) extends InputValue 
  case class SimpleValue(name: String)                    extends InputValue
}
```
And it was called like this:
```scala
implicit lazy val inputValueSchema: Schema[Any, InputValue] = Schema.derived
```
During the derivation will need an instance of itself because it's recursive, but it was finding itself thanks to `lazy`.

**Comment by odersky** [comment]
@ghostdogpr Can we minimize this to a self-contained example? That would be important to see whether the code is actually safe and, if that's the case, to refine the checking analysis.

**Comment by ghostdogpr** [comment]
> @ghostdogpr Can we minimize this to a self-contained example? That would be important to see whether the code is actually safe and, if that's the case, to refine the checking analysis.

Here's a minimized example: https://gist.github.com/ghostdogpr/b525ee4062427fa06861b96e3544fae2
It works well with `3.1.1` and the warning shows up with `3.2.1-RC1-bin-20220705-9bb3108-NIGHTLY`.
Hope it helps!

**Comment by odersky** [comment]
@ghostdogpr Thanks for the minimization! It seems the change in this PR uncovered another bug in the loop checking. Fix forthcoming.

**Pull request merged**